### PR TITLE
Optimize Memory Handling in Embedding Computations and Refactor EmbeddingService

### PR DESCRIPTION
**Pull request by heispv**
### **Description:**
This PR enhances the `EmbeddingService` by optimizing memory usage during embedding computations, refactoring core logic, and adding unit tests for robustness.

### **Key Changes:**
1. **Dynamic Memory Management:**
   - Automatically estimate and manage RAM usage by saving embeddings to disk when memory limits are reached.

2. **Refactor `EmbeddingService`:**
   - Extracted core logic into a dedicated method.
   - Handled ultra-long reads by saving embeddings immediately and preventing additional sequence loading.
   - Removed `SAVE_AFTER_N_EMBEDDINGS`.

3. **Unit Testing:**
   - Added tests for computing embeddings on long, short, and mixed sequences.
   - Validated dynamic memory management and embedding computation accuracy.
   - Ensured correct result paths and file existence in the tests.

### **Issue Reference:**
Closes [#98]

### CODE REVIEW

**Review by SebieF — CHANGES_REQUESTED** [review]
Very nice work, thank you for looking into this quite complex problem! :)

Most of my comments are minor suggestions for improvements, the only major change that I would suggest is to unify the handling of the ultra-long reads and normal reads if at all possible, and removing the while loops altogether. 

Other than that, the approach seems to work fine, I ran the tests locally and executed a run with ProtT5 for 2377 sequences successfully.

**Comment by heispv** [comment]
Thank you for the suggestions on improving the changes I made. I just updated the code based on your feedback! :)

File: biotrainer/embedders/embedding_service.py [hunk 1]
```diff
@@ -68,45 +75,146 @@ def compute_embeddings(self, sequence_file: str, output_dir: Path, protocol: Pro
...
+        max_embedding_fit = self._max_embedding_fit(embeddings[sequence_ids[0]])
+        
+        if embeddings[sequence_ids[0]] is None:
+            logger.info(f"No embeddings found.")
+            return str(embeddings_file_path)
+        
+        logger.info(f"Embedding dimension: {embeddings[sequence_ids[idx]].shape[-1]}")
+        logger.info("If your dataset contains long reads, it may take more time to process the first few sequences.")
+        
+        with tqdm(total=total_sequences, initial=1, desc="Processing Sequences") as pbar:
```
Comment (SebieF) [inline biotrainer/embedders/embedding_service.py:118]: I just checked locally, and the processing bar is not shown yet for the first embedding calculation. Could you move the whole line to line 108, then it should be shown. You have to double check if the initial value is still correct (0 or 1?) and probably have to call `pbar.update(1)` once manually after the line 108. But I think the improved user feedback is worth this small overhead.

Additionally, I would suggest to use `desc="Computing embeddings"` to be clearer about the description (but adding the description is a good idea anyway, thanks!)

File: biotrainer/embedders/embedding_service.py [hunk 2]
```diff
@@ -68,45 +75,171 @@ def compute_embeddings(self, sequence_file: str, output_dir: Path, protocol: Pro
...
+            last_save_id, embeddings = self._save_and_reset_embeddings(embeddings, idx, last_save_id,
+                                                            embeddings_file_path, use_reduced_embeddings)
 
-        # Delete embeddings and embedding model from memory now, because they will no longer be needed
+        end_time = time.time()
+        logger.info(f"Time elapsed for saving embeddings: {end_time - start_time:.2f}[s]")
+        
         del embeddings
-        del self._embedder
         gc.collect()
```
[the remark this answers is not in the document]
Comment (SebieF) [reply biotrainer/embedders/embedding_service.py:149]: Sorry I think I did not express myself well here, please add the `del self._embedder` statement here and remove the `gc.collect()` call from line 81, such that the embedding model is deleted, but the garbage collector is only called once.

Lines 148-150 should look like this then:
```python
        del embeddings 
        del self._embedder
        gc.collect()
```

File: biotrainer/embedders/embedding_service.py [hunk 3]
```diff
@@ -68,45 +75,146 @@ def compute_embeddings(self, sequence_file: str, output_dir: Path, protocol: Pro
...
 
         return str(embeddings_file_path)
 
+    @staticmethod
+    def _max_embedding_fit(embedding: ndarray) -> int:
+        """
+        Calculates the maximum number of embeddings that can fit in available memory.
```
Comment (SebieF) [inline biotrainer/embedders/embedding_service.py:156]: Please add an explanation to the doc string how you found the magic constants in this function (`0.75` and `18`) and how they can be interpreted

File: tests/test_embedding_service.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        self._setup_fasta_parameters()
+
+    def _setup_test_environment(self):
+        logger.info('Setting up test environment')
+        self.embedder = OneHotEncodingEmbedder()
+        self.embedding_service = EmbeddingService(embedder=self.embedder)
+
+    def _setup_fasta_parameters(self):
+        self.num_reads = 100
+        if os.environ.get('CI') == 'true':
```
Comment (SebieF) [inline tests/test_embedding_service.py:39]: Nice solution, and seems to work as the tests took about 11mins. Did you double check that it works in the CI pipeline though? Might be nice to put a logging statement here to be able to make it easier understandable in the pipeline, why the test takes so long (for new developers for example).

**Review by SebieF — CHANGES_REQUESTED** [review]
Thanks, the embedding calculation is much easier to understand now with your solution to use only one loop! :+1:  

I only have a few small suggestions left, then we are ready to rebase. :smile:

**Comment by heispv** [comment]
Sebastian, I just updated the code as you suggested 😉

**Review by SebieF — APPROVED** [review]
Thank you very much! :)

**Pull request merged**